lib/jsonWebToken.js

```diff
@@ -0,0 +1,195 @@
+      callback.apply(null, args);
+    });
+  };
+
+  // make sure we have the required fields in options
+  if (typeof options.audience !== 'string' || options.audience === '')
+    return done(new Error('options.audience is missing in jsonWebToken.verify'));
+  if (!options.algorithms)
+    return done(new Error('options.algorithms is missing in jsonWebToken.verify'));
+  if (!Array.isArray(options.algorithms) || options.algorithms.length == 0 ||
+    (options.algorithms.length == 1 && options.algorithms[0] === 'none'))
+    return done(new Error('options.algorithms must be an array containing at least one algorithm'));
+
+  /*********************************************************************
+   * Checking jwtString structure, getting header, payload and signature
+   ********************************************************************/
+
+  // split jwtString, make sure we have three parts and we have signature
+  var parts = jwtString.split('.');
+  if (parts.length !== 3)
+    return done(new Error('jwtString is malformed'));
+  if (parts[2] === '')
+    return done(new Error('signature is missing in jwtString'));
+  
+  // decode jwsString
+  var decodedToken;
+  try {
+    decodedToken = jws.decode(jwtString);
+  } catch(err) {
+    return done(new Error('invalid token'));
+  }
+
+  if (!decodedToken) {
+    return done(new Error('invalid token'));
+  }
+
+  // get header, payload and signature
+  var header = decodedToken.header;
+  var payload = decodedToken.payload;
+  var signature = decodedToken.signature;
+
+  if (!header)
+    return done(new Error('missing header in the token'));
+  if (!payload)
+    return done(new Error('missing payload in the token'));
+  if (!signature)
+    return done(new Error('missing signature in the token'));
+
+  /*********************************************************************
+   * validate algorithm and signature
+   ********************************************************************/
+
+  // header.alg should be one of the algorithms provided in options.algorithms
+  if (typeof header.alg !== 'string' || header.alg === '' || header.alg === 'none' ||
+    options.algorithms.indexOf(header.alg) == -1) {
+    return done(new Error('invalid algorithm'));
+  }
+
+  try {
+    var valid = jws.verify(jwtString, header.alg, PEMKey);
+    if (!valid)
+      return done(new Error('invalid signature'));
+  } catch (e) {
+    return done(e);
+  }
+
+  /*********************************************************************
+   * validate payload content
+   ********************************************************************/
+
+  // (1) issuer
+  //   - check the existence and the format of payload.iss
+  //   - validate if options.issuer is set
+  if (typeof payload.iss !== 'string' || payload.iss === '')
+    return done(new Error('invalid iss value in payload'));
+  if (options.issuer) {
+    if ((Array.isArray(options.issuer) && options.issuer.indexOf(payload.iss) == -1) ||
```

Comment:
@brentschmaltz Can you comment on this? To support multiple issuers, we are support options.issuer as an Array, but would payload.iss come back as an array of issuers as well if we're going against the common metadata endpoint?
@lovemaths It might be worth just trying it. What does payload.iss return when you go against common?